Sources/SourceKitLSP/Swift/SyntaxTreeManager.swift

```diff
@@ -0,0 +1,112 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftParser
+import SwiftSyntax
+
+/// Keeps track of SwiftSyntax trees for document snapshots and computes the
+/// SwiftSyntax trees on demand.
```

Comment:
Is the thought that the highlight ranges in the semantic case aren't as expensive as the entire syntax tree? ie. why does one have a cache limit but the other doesn't? They're also both quite similar, ie. take an edit and cache the result (which here is the syntax tree and in the semantic is the highlight ranges).

Reply:
The difference is that we can re-compute the syntax tree on demand but we can’t (in the current design) re-compute the semantic tokens. That’s why we can be more aggressive about discarding syntax trees than semantic tokens. If we wanted to change that, getting semantic tokens would need to become a sourcekitd request as well, instead of being sent over implicitly on every update (which we should probably do).